text/0000-singleton-removal.md

```diff
@@ -0,0 +1,69 @@
+
+A few things have been left static. `BitWriterPool` doesn't make use of `NetworkingManager` at all, and has consequently been left shared. The `LogLevel` has also been left static, as it is a bother to thread through into some of the places where it is used (however, I can see the argument for biting the bullet and doing so, as it would be nice to see whether log lines were generated from the server or client, respectively).
+
+For editor code, there is a new static method: `NetworkingManagerEditor.GetAnyNetworkingManager` that just wraps a [`FindObjectOfType`](https://docs.unity3d.com/ScriptReference/Object.FindObjectOfType.html) call. This is the replacement for editor code that needed to operate on "the NetworkingManager" (back when there was only one). 
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Singletons are convenient. For code that operates in the context of a `NetworkedObject`, this isn't a loss, but for any other code, it requires more thought. For games that have multiple `NetworkingManager`s (particularly in the same scene), there is no recourse to making users think harder about _which_ `NetworkingManager` they want to interact with. However, for games that don't make use of multiple `NetworkingManager`s, it creates a pain point with no upside. I've considered if we should just leave the `NetworkingManager.Singleton` in place (with the first `NetworkingManager` to claim it "winning"), just to maintain continuity for users that don't need this feature. However, there is a risk of this creating bugs and confusion for users who do start down the path of multiple `NetworkingManager`s.
+
+This also makes the editor facilities feel a bit more halfbaked. It would be nice if the MLAPIProfiler could provide feedback for all the `NetworkingManager` instances that are running, but this will entail more work. 
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+There is no alternative design that covers all use cases of removing MLAPI singleton dependence. An architecturally similar approach to supporting host-based games is to have the client start up a separate, out-of-process server, but this will increase the minspec of the Host machine, because it effectively doubles its memory requirements. There is no good solution to talking to multiple servers simultaneously via MLAPI. It may be possible to live without this feature when writing MMOs, as a common pattern is to have the client only talk to a single gateway server in production; however, forcing this pattern onto development is undesirable, as configuring a local gateway server can be nontrivial. 
+
+# Prior art
+[prior-art]: #prior-art
+
+There is no specific prior art to cite, but avoiding static global state is a common software development recommendation, as it reduces system isolation and flexibility.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
```

Comment:
What about `NetworkingManager.NetworkTime`?  Do we like each NetworkingManager having it's own un-synchronized copy (each with its own currentNetworkTimeOffset)?

Reply:
I think all NetworkingManagers need their own NetworkTime, that synchronize with their server time in the same way as before (in the case of inprocess server/client, the latency offset will of course be close to 0!)